kythe/rust/indexer/src/providers.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::KytheError;
+use analysis_rust_proto::*;
+use positioned_io;
+use rc_zip::{prelude::*, Archive, EntryContents};
+use std::fs::File;
+use std::io::Read;
+
+/// A trait for retrieving files during indexing.
+pub trait FileProvider {
+  /// Checks whether a file exists and that the contents is available.
+  ///
+  /// This function takes a file hash and returns whether
+  /// the file exists and that the contents can be requested.
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a [KytheError::KzipFileError] will
+  /// be returned.
+  fn exists(&mut self, file_hash: &str) -> bool;
+
+  /// Retrieves the string contents of a file.
+  ///
+  /// This function takes a file hash and returns the string contents of the
+  /// file.
+  ///
+  /// # Errors
+  ///
+  /// If the file does not exist, a
+  /// [FileNotFoundError][KytheError::FileNotFoundError] will be returned.
+  /// If the file can't be read into a String, a
+  /// [FileReadError][KytheError::FileReadError] will be returned.
+  fn contents(&mut self, file_hash: &str) -> Result<String, KytheError>;
+}
+
+/// A [FileProvider] that backed by a .kzip file.
+pub struct KzipFileProvider {
+  file: File,
+  zip_archive: Archive,
+}
+
+impl KzipFileProvider {
+  /// Create a new instance using a valid [File] object of a kzip
+  /// file.KytheError
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a
+  /// [KzipFileError][KytheError::KzipFileError] will be returned.
+  pub fn new(file: File) -> Result<Self, KytheError> {
+    let zip_archive = match file.read_zip() {
+      Ok(zip_archive) => zip_archive,
+      Err(_e) => return Err(KytheError::KzipFileError),
```

Comment:
Nitpick, but I prefer just `Err(_)` here and elsewhere. The `e` doesn't add much.

Also, consider using [`.map_err(|_| KytheError::KzipFileError)?`](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) instead.

Reply:
Gotcha. I'm going to be updating the error struct to support additional traits so this code will be changed to support that new functionality.